# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')    
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id   
0          1   -8437.673028         42               высшее             0  \
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt   
0    женат / замужем                 0      F   сотрудник     0  \
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [14]:
data.duplicated().sum()

54

In [15]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

`Рассмотрим датафреймы 'debt_availability' и 'no_debt_availability' с наличием и отсутствием задолженности у клиентов для дальнейшего анализа:`

In [22]:
debt_availability = data[data['debt'] == 1] # датафрейм с наличием задолженности у клиентов
no_debt_availability = data[data['debt'] != 1] # # датафрейм с отсутствием задолженности у клиентов

`Рассмотрим процентное соотношение клиентов по задолженностям:`

In [23]:
percent_debt_availability = debt_availability['debt'].count() / data['debt'].count()
percent_no_debt_availability = no_debt_availability['debt'].count() / data['debt'].count()
print(f'Общий процент по наличию задолженностей среди клиентов: {round(percent_debt_availability, 3) * 100} %')
print(f'Общий процент по отсутствию задолженностей среди клиентов: {round(percent_no_debt_availability, 3) * 100} %')

Общий процент по наличию задолженностей среди клиентов: 8.1 %
Общий процент по отсутствию задолженностей среди клиентов: 91.9 %


`Разберемся с тем, как наличие и количество детей влияет на возврат кредита в срок:`

In [24]:
children_debt = debt_availability.groupby('children')['debt'].count().reset_index()
children_no_debt = no_debt_availability.groupby('children')['debt'].count().reset_index()

#добавим строку с нулевым значением, так как строка важна для дальнейшего анализа:
#row = [{'children': 5, 'debt': 0}] 
#row = pd.DataFrame(row)
#children_debt.loc[5] = row#children_debt.append(row, ignore_index=True)
#переименование столбца для визуального упрощения:
children_no_debt.rename(columns={'debt': 'no_debt'}, inplace=True)
display(children_debt)
display(children_no_debt)

children  debt
0         0  1063
1         1   444
2         2   194
3         3    27
4         4     4

children  no_debt
0         0    13044
1         1     4365
2         2     1858
3         3      303
4         4       37
5         5        9

`Из чего можно сделать вывод:`

In [26]:
absolut_value_child = children_debt['debt'] / children_no_debt['no_debt']
for i in range(len(children_debt)):
    absolut_value_child[i] = children_debt['debt'][i] / children_no_debt['no_debt'][i]
    print(f'У клиента с {i} детьми вероятность возникновение задолженности: {round(absolut_value_child[i], 3) * 100} %')
    

У клиента с 0 детьми вероятность возникновение задолженности: 8.1 %
У клиента с 1 детьми вероятность возникновение задолженности: 10.2 %
У клиента с 2 детьми вероятность возникновение задолженности: 10.4 %
У клиента с 3 детьми вероятность возникновение задолженности: 8.9 %
У клиента с 4 детьми вероятность возникновение задолженности: 10.8 %


**Вывод:** 

По полученным данным, можно сказать следующее: Зависимости по наличию задолженности у клиентов и количеству их детей внутри категории незначительна. Исключение составляет только клиенты с 5 детьми, так как их выборка крайне мала (9 человек).

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [27]:
family_status_debt = debt_availability.groupby('family_status')['debt'].count().reset_index()
family_status_no_debt = no_debt_availability.groupby('family_status')['debt'].count().reset_index()
family_status_no_debt.rename(columns={'debt': 'no_debt'}, inplace=True)

In [28]:
absolut_value_family_status = family_status_debt['debt'] / family_status_no_debt['no_debt']
status = family_status_debt['family_status']
for i in range(len(status)):
    absolut_value_family_status[i] = family_status_debt['debt'][i] / family_status_no_debt['no_debt'][i]
    print(f'У клиента, который {status[i]} вероятность возникновение задолженности: {round(absolut_value_family_status[i], 5) * 100} %')

У клиента, который Не женат / не замужем вероятность возникновение задолженности: 10.82 %
У клиента, который в разводе вероятность возникновение задолженности: 7.602 %
У клиента, который вдовец / вдова вероятность возникновение задолженности: 7.095 %
У клиента, который гражданский брак вероятность возникновение задолженности: 10.237 %
У клиента, который женат / замужем вероятность возникновение задолженности: 8.175 %


**Вывод:** 

Такое семейное положение клиента как "Не женат / не замужем" и "гражданский брак" выделяются от остальных категорий не в лучшую сторону.

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [29]:
total_income_category_debt = debt_availability.groupby('total_income_category')['debt'].count().reset_index()
total_income_category_no_debt = no_debt_availability.groupby('total_income_category')['debt'].count().reset_index()
total_income_category_no_debt.rename(columns={'debt': 'no_debt'}, inplace=True)

In [30]:
absolut_value_category = total_income_category_debt['debt'] / total_income_category_no_debt['no_debt']
category = total_income_category_debt['total_income_category']
for i in range(len(category)):
    absolut_value_category[i] = total_income_category_debt['debt'][i] / total_income_category_no_debt['no_debt'][i]
    print(f'У клиента, который относится к типу {category[i]} по доходам вероятность возникновение задолженности: {round(absolut_value_category[i], 4) * 100} %')

У клиента, который относится к типу A по доходам вероятность возникновение задолженности: 8.7 %
У клиента, который относится к типу B по доходам вероятность возникновение задолженности: 7.6 %
У клиента, который относится к типу C по доходам вероятность возникновение задолженности: 9.28 %
У клиента, который относится к типу D по доходам вероятность возникновение задолженности: 6.4 %
У клиента, который относится к типу E по доходам вероятность возникновение задолженности: 10.0 %


**Вывод:** 

Клиенты категории С по доходам оказались самыми ненадежными клиентами по наличию задолженности. Клиентов категории Е рассматривать неуместно, их выборка мала. 

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

In [31]:
purpose_category_debt = debt_availability.groupby('purpose_category')['debt'].count().reset_index()
purpose_category_no_debt = no_debt_availability.groupby('purpose_category')['debt'].count().reset_index()
purpose_category_no_debt.rename(columns={'debt': 'no_debt'}, inplace=True)

In [32]:
absolut_value_category = purpose_category_debt['debt'] / purpose_category_no_debt['no_debt']
category = purpose_category_debt['purpose_category']
for i in range(len(category)):
    absolut_value_category[i] = purpose_category_debt['debt'][i] / purpose_category_no_debt['no_debt'][i]
    print(f'У клиента с целью {category[i]} на кредитные деньги вероятность возникновение задолженности: {round(absolut_value_category[i], 4) * 100} %')

У клиента с целью операции с автомобилем на кредитные деньги вероятность возникновение задолженности: 10.31 %
У клиента с целью операции с недвижимостью на кредитные деньги вероятность возникновение задолженности: 7.82 %
У клиента с целью получение образования на кредитные деньги вероятность возникновение задолженности: 10.190000000000001 %
У клиента с целью проведение свадьбы на кредитные деньги вероятность возникновение задолженности: 8.55 %


**Вывод:** 

Операции с авто и получение образования - категории, немного вырывающиеся вперед по вероятности появления задолженности у клиента.

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

причины появления пропусков в исходных данных можно разделить на 2 типа:
- технический
- человеческий фактор

К первому типу можно отнести моменты, связанные с переносами датафреймов с одного сервера на другой, или же просто внутри директории. При таких действиях данные могут "потеряться". Так же несовершенство некоторых техпроцессов  могут привести к появлению пропусков (данные могут попросту потеряться по пути от клиента браузера до сервера).

Человеческий фактр так же оказывает большое влияние на наличие пропусков, ведь большое количество данных заполняется сотрудниками организации или клиентами этих организаций вручную. 

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

В базах данных встречаются аномальные значения(резко малые или резко большие), во избежания влияния их на результат анализа данных, стоит использовать медиану.
(как по мне так лучше моду, но в анализе слишком большое разнообразие количественнных значений)

## Общий вывод.

Общий процент по наличию задолженностей среди клиентов:   8.1 %,

Вероятность появления задолженности в зависимости:

                                - количеству детей:    8.1 - 10.8 %,                                
                                - семейное положение:  7.1 - 10.82 %,                               
                                - уровень дохода:      6.4 - 10 %,                                
                                - цель кредита:        7.8 - 10.3 %.                              
Разброс в отдельно взятой категории незначительный. Но по всем категориям можно выделить потенциально нежелательного клиента для банка.

Взять к примеру такого клиента: вдовца без детей, берущий кредитные средства на операциии с недвижимостью. 



In [33]:
positiv_client_probability = data[data['children'] == 0]
positiv_client_probability = positiv_client_probability[positiv_client_probability['family_status'] == 'вдовец / вдова']
positiv_client_probability = positiv_client_probability[positiv_client_probability['purpose_category'] == 'операции с недвижимостью']
positiv_client_probability_count = positiv_client_probability['debt'].count()
positiv_client_probability_debt = positiv_client_probability[positiv_client_probability['debt'] == 1]['debt'].count() 
print(f'Таких клиентов в банке: {positiv_client_probability_count}, с задолженностью из них - {positiv_client_probability_debt}, что равно {positiv_client_probability_debt / positiv_client_probability_count * 100} %')

Таких клиентов в банке: 470, с задолженностью из них - 22, что равно 4.680851063829787 %


Рассмотрим клиента по следующим характеристикам: 1 ребенок, не женат, цель кредита - приобретение автомобиля.

In [34]:
negative_client_probability = data[data['children'] == 1]
negative_client_probability = negative_client_probability[negative_client_probability['family_status'] == 'Не женат / не замужем']
negative_client_probability = negative_client_probability[negative_client_probability['purpose_category'] == 'операции с автомобилем']
negative_client_probability_count = negative_client_probability['debt'].count()
negative_client_probability_debt = negative_client_probability[negative_client_probability['debt'] == 1]['debt'].count() 
print(f'Таких клиентов в банке: {negative_client_probability_count}, с задолженностью из них - {negative_client_probability_debt}, что равно {negative_client_probability_debt / negative_client_probability_count * 100} %')

Таких клиентов в банке: 93, с задолженностью из них - 13, что равно 13.978494623655912 %


Таким образом, имея полученные данные и применяя сортировку по категориям, банк может иметь представление о клиенте и принять дальнейшее решение как с ним стоит работать.